codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -139,8 +161,36 @@ public Http2Stream answer(InvocationOnMock invocation) throws Throwable {
         when(writer.writeSettings(eq(ctx), any(Http2Settings.class), eq(promise))).thenReturn(future);
         when(writer.writeGoAway(eq(ctx), anyInt(), anyInt(), any(ByteBuf.class), eq(promise)))
                 .thenReturn(future);
-        when(remoteFlow.sendFlowControlledFrame(eq(ctx), any(Http2Stream.class),
-                 any(ByteBuf.class), anyInt(), anyBoolean(), eq(promise))).thenReturn(future);
+        writtenData = new ArrayList<String>();
+        writtenPadding = new ArrayList<Integer>();
+        final AtomicBoolean streamClosed = new AtomicBoolean();
+        when(writer.writeData(eq(ctx), anyInt(), any(ByteBuf.class), anyInt(), anyBoolean(), any(ChannelPromise.class)))
+                .then(new Answer<Object>() {
+            @Override
+            public Object answer(InvocationOnMock invocationOnMock) throws Throwable {
+                // Make sure we only receive stream closure on the last frame and that void promises are used for
+                // all writes except the last one.
+                ChannelPromise receivedPromise = (ChannelPromise) invocationOnMock.getArguments()[5];
+                if (streamClosed.get()) {
+                    fail("Stream already closed");
+                } else {
+                    streamClosed.set((Boolean) invocationOnMock.getArguments()[4]);
+                    if (streamClosed.get()) {
+                        assertSame(promise, receivedPromise);
+                    }
+                }
+                writtenPadding.add((Integer) invocationOnMock.getArguments()[3]);
+                ByteBuf data = (ByteBuf) invocationOnMock.getArguments()[2];
+                writtenData.add(data.toString(UTF_8));
+                data.skipBytes(data.readableBytes());
```

Comment:
Can you add some brief comments why we need to skip bytes and then release the buffer?